File: contest/LeetCode_Weekly_Contest_206/Check_If_String_Is_Transformable_With_Substring_Sort_Operations.js
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isTransformable = function(s, t) {
    var pos = [], top = []
    for (var i = 0; i < 10; i++)
        pos[i] = [], top[i] = 0
    
    var n = s.length
    for (var i = 0; i < n; i++)
        pos[s[i] - '0'].push(i)
    
    var m = t.length
    for (var i = 0; i < m; i++) {
        var idx = t[i] - '0'
        if (top[idx] >= pos[idx].length) return false
        
        for (var j = 0; j < idx; j++)
            if (top[j] < pos[j].length && pos[j][top[j]] < pos[idx][top[idx]])
                return false
        top[idx]++
    }
    return true
};

File: contest/LeetCode_Weekly_Contest_206/Check_If_String_Is_Transformable_With_Substring_Sort_Operations.java
class Solution {
    ArrayList<Integer>[] pos;
    int n;
    public boolean isTransformable(String s, String t) {
        n = s.length();

        pos = new ArrayList[10];
        for (int i = 0; i < 10; ++i)
            pos[i] = new ArrayList<Integer>();

        for (int i = 0; i < n; ++i) {
            int idx = s.charAt(i) - '0';
            pos[idx].add(i);
        }

        int[] p = new int[10];
        for (int i = 0; i < 10; ++i) p[i] = 0;

        for (int i = 0; i < t.length(); ++i) {
            int idx = t.charAt(i) - '0';
            if (p[idx] >= pos[idx].size()) return false;

            int where = pos[idx].get(p[idx]);
            p[idx]++;

            for (int v = 0; v < idx; ++v) {
                if (p[v] >= pos[v].size()) continue;
                if (pos[v].get(p[v]) < where) return false;
            }
        }

        return true;
    }
}
